File: lib/core/services/web3/ipfs_service.dart
/// GramPulse IPFS Service
///
/// Handles uploading proof files to IPFS via the attestation backend.
/// This service is used when resolving grievances to store immutable
/// proof of the resolution.

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';
import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';
import '../config/web3_config.dart';

/// Result of an IPFS upload
class IPFSUploadResult {
  final bool success;
  final String? cid;
  final String? gatewayUrl;
  final String? error;
  final int? size;

  IPFSUploadResult({
    required this.success,
    this.cid,
    this.gatewayUrl,
    this.error,
    this.size,
  });

  factory IPFSUploadResult.fromJson(Map<String, dynamic> json) {
    if (json['success'] == true) {
      final data = json['data'] as Map<String, dynamic>;
      return IPFSUploadResult(
        success: true,
        cid: data['cid'] as String?,
        gatewayUrl: data['gatewayUrl'] as String?,
        size: data['size'] as int?,
      );
    } else {
      return IPFSUploadResult(
        success: false,
        error: json['message'] as String? ?? 'Unknown error',
      );
    }
  }
}

/// Result of a proof package upload
class ProofPackageResult {
  final bool success;
  final String? packageCid;
  final String? packageUrl;
  final List<MediaFileResult> mediaFiles;
  final String? error;

  ProofPackageResult({
    required this.success,
    this.packageCid,
    this.packageUrl,
    this.mediaFiles = const [],
    this.error,
  });

  factory ProofPackageResult.fromJson(Map<String, dynamic> json) {
    if (json['success'] == true) {
      final data = json['data'] as Map<String, dynamic>;
      final mediaList = (data['mediaFiles'] as List<dynamic>?)
          ?.map((m) => MediaFileResult.fromJson(m as Map<String, dynamic>))
          .toList() ?? [];
      
      return ProofPackageResult(
        success: true,
        packageCid: data['packageCid'] as String?,
        packageUrl: data['packageUrl'] as String?,
        mediaFiles: mediaList,
      );
    } else {
      return ProofPackageResult(
        success: false,
        error: json['message'] as String? ?? 'Unknown error',
      );
    }
  }
}

/// Individual media file result
class MediaFileResult {
  final String cid;
  final String fileName;
  final String mimeType;
  final String gatewayUrl;

  MediaFileResult({
    required this.cid,
    required this.fileName,
    required this.mimeType,
    required this.gatewayUrl,
  });

  factory MediaFileResult.fromJson(Map<String, dynamic> json) {
    return MediaFileResult(
      cid: json['cid'] as String,
      fileName: json['fileName'] as String,
      mimeType: json['mimeType'] as String,
      gatewayUrl: json['gatewayUrl'] as String,
    );
  }
}

/// IPFS Service for GramPulse
///
/// Provides methods to upload proof files to IPFS through the backend
class IPFSService {
  static IPFSService? _instance;
  
  final String _baseUrl;
  final String _apiKey;
  final http.Client _client;

  IPFSService._({
    required String baseUrl,
    required String apiKey,
    http.Client? client,
  })  : _baseUrl = baseUrl,
        _apiKey = apiKey,
        _client = client ?? http.Client();

  /// Get singleton instance
  static IPFSService get instance {
    if (_instance == null) {
      final config = Web3Config.instance;
      _instance = IPFSService._(
        baseUrl: config.attestationServiceUrl,
        apiKey: config.apiKey,
      );
    }
    return _instance!;
  }

  /// Initialize with custom configuration (for testing)
  static void initialize({
    required String baseUrl,
    required String apiKey,
    http.Client? client,
  }) {
    _instance = IPFSService._(
      baseUrl: baseUrl,
      apiKey: apiKey,
      client: client,
    );
  }

  /// Reset instance (for testing)
  static void reset() {
    _instance = null;
  }

  /// Upload a single file to IPFS
  ///
  /// [file] - The file to upload
  /// [grievanceId] - Associated grievance ID (optional)
  ///
  /// Returns [IPFSUploadResult] with CID and gateway URL
  Future<IPFSUploadResult> uploadFile(
    File file, {
    String? grievanceId,
  }) async {
    try {
      final uri = Uri.parse('$_baseUrl/ipfs/upload');
      final request = http.MultipartRequest('POST', uri);

      // Add headers
      request.headers['x-api-key'] = _apiKey;

      // Add file
      final fileName = file.path.split(Platform.pathSeparator).last;
      final mimeType = _getMimeType(fileName);
      
      request.files.add(await http.MultipartFile.fromPath(
        'file',
        file.path,
        contentType: MediaType.parse(mimeType),
      ));

      // Add grievance ID if provided
      if (grievanceId != null) {
        request.fields['grievanceId'] = grievanceId;
      }

      // Send request
      final streamedResponse = await _client.send(request);
      final response = await http.Response.fromStream(streamedResponse);

      if (response.statusCode == 201) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return IPFSUploadResult.fromJson(json);
      } else {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return IPFSUploadResult(
          success: false,
          error: json['message'] as String? ?? 'Upload failed: ${response.statusCode}',
        );
      }
    } catch (e) {
      return IPFSUploadResult(
        success: false,
        error: 'Upload error: $e',
      );
    }
  }

  /// Upload file from bytes
  Future<IPFSUploadResult> uploadBytes(
    Uint8List bytes,
    String fileName, {
    String? grievanceId,
  }) async {
    try {
      final uri = Uri.parse('$_baseUrl/ipfs/upload');
      final request = http.MultipartRequest('POST', uri);

      // Add headers
      request.headers['x-api-key'] = _apiKey;

      // Add file
      final mimeType = _getMimeType(fileName);
      request.files.add(http.MultipartFile.fromBytes(
        'file',
        bytes,
        filename: fileName,
        contentType: MediaType.parse(mimeType),
      ));

      // Add grievance ID if provided
      if (grievanceId != null) {
        request.fields['grievanceId'] = grievanceId;
      }

      // Send request
      final streamedResponse = await _client.send(request);
      final response = await http.Response.fromStream(streamedResponse);

      if (response.statusCode == 201) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return IPFSUploadResult.fromJson(json);
      } else {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return IPFSUploadResult(
          success: false,
          error: json['message'] as String? ?? 'Upload failed: ${response.statusCode}',
        );
      }
    } catch (e) {
      return IPFSUploadResult(
        success: false,
        error: 'Upload error: $e',
      );
    }
  }

  /// Create a complete proof-of-resolution package
  ///
  /// This uploads all proof files and creates a metadata package on IPFS
  ///
  /// [grievanceId] - Firebase document ID of the grievance
  /// [villageId] - Village identifier
  /// [resolverRole] - Role of the resolver (officer/volunteer)
  /// [resolverId] - User ID of the resolver
  /// [description] - Resolution description
  /// [files] - List of proof files to upload
  ///
  /// Returns [ProofPackageResult] with package CID
  Future<ProofPackageResult> createProofPackage({
    required String grievanceId,
    required String villageId,
    required String resolverRole,
    required String resolverId,
    String? description,
    List<File>? files,
  }) async {
    try {
      final uri = Uri.parse('$_baseUrl/ipfs/proof-package');
      final request = http.MultipartRequest('POST', uri);

      // Add headers
      request.headers['x-api-key'] = _apiKey;

      // Add fields
      request.fields['grievanceId'] = grievanceId;
      request.fields['villageId'] = villageId;
      request.fields['resolverRole'] = resolverRole;
      request.fields['resolverId'] = resolverId;
      if (description != null) {
        request.fields['description'] = description;
      }

      // Add files
      if (files != null) {
        for (final file in files) {
          final fileName = file.path.split(Platform.pathSeparator).last;
          final mimeType = _getMimeType(fileName);
          
          request.files.add(await http.MultipartFile.fromPath(
            'files',
            file.path,
            contentType: MediaType.parse(mimeType),
          ));
        }
      }

      // Send request
      final streamedResponse = await _client.send(request);
      final response = await http.Response.fromStream(streamedResponse);

      if (response.statusCode == 201) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return ProofPackageResult.fromJson(json);
      } else {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return ProofPackageResult(
          success: false,
          error: json['message'] as String? ?? 'Package creation failed: ${response.statusCode}',
        );
      }
    } catch (e) {
      return ProofPackageResult(
        success: false,
        error: 'Package error: $e',
      );
    }
  }

  /// Get gateway URL for a CID
  String getGatewayUrl(String cid) {
    // Use Pinata gateway by default
    return 'https://gateway.pinata.cloud/ipfs/$cid';
  }

  /// Check if IPFS service is available
  Future<bool> isAvailable() async {
    try {
      final uri = Uri.parse('$_baseUrl/ipfs/health');
      final response = await _client.get(uri);
      
      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return json['status'] == 'healthy';
      }
      return false;
    } catch (e) {
      return false;
    }
  }

  /// Get MIME type from file extension
  String _getMimeType(String fileName) {
    final extension = fileName.split('.').last.toLowerCase();
    switch (extension) {
      case 'jpg':
      case 'jpeg':
        return 'image/jpeg';
      case 'png':
        return 'image/png';
      case 'webp':
        return 'image/webp';
      case 'mp4':
        return 'video/mp4';
      case 'webm':
        return 'video/webm';
      default:
        return 'application/octet-stream';
    }
  }
}

File: backend/src/services/ipfsService.js
/**
 * GramPulse IPFS Service
 * 
 * Handles IPFS uploads for proof-of-resolution files
 * Uses Pinata or Web3.Storage for pinning
 */

const axios = require('axios');
const FormData = require('form-data');
const { v4: uuidv4 } = require('uuid');
const logger = require('../utils/logger');

/**
 * IPFS Service Configuration
 */
const IPFS_CONFIG = {
  pinata: {
    apiUrl: 'https://api.pinata.cloud',
    gateway: 'https://gateway.pinata.cloud/ipfs',
  },
  web3Storage: {
    apiUrl: 'https://api.web3.storage',
    gateway: 'https://w3s.link/ipfs',
  },
};

class IPFSService {
  constructor() {
    this.provider = process.env.IPFS_PROVIDER || 'pinata';
    this.apiKey = process.env.IPFS_API_KEY;
    this.apiSecret = process.env.IPFS_API_SECRET;
    this.initialized = false;
  }

  /**
   * Initialize the IPFS service
   */
  async initialize() {
    if (!this.apiKey) {
      logger.warn('IPFS API key not configured. IPFS uploads will be disabled.');
      return;
    }

    try {
      // Test connection based on provider
      if (this.provider === 'pinata') {
        await this.testPinataConnection();
      }
      
      this.initialized = true;
      logger.info(`IPFS Service initialized with provider: ${this.provider}`);
    } catch (error) {
      logger.error('Failed to initialize IPFS Service:', error);
      throw error;
    }
  }

  /**
   * Test Pinata API connection
   */
  async testPinataConnection() {
    const response = await axios.get(
      `${IPFS_CONFIG.pinata.apiUrl}/data/testAuthentication`,
      {
        headers: {
          'pinata_api_key': this.apiKey,
          'pinata_secret_api_key': this.apiSecret,
        },
      }
    );
    
    if (response.status !== 200) {
      throw new Error('Pinata authentication failed');
    }
    
    logger.info('Pinata connection verified');
  }

  /**
   * Upload a file buffer to IPFS
   * 
   * @param {Buffer} fileBuffer - File content as buffer
   * @param {string} fileName - Original filename
   * @param {Object} metadata - Additional metadata
   * @returns {Object} - Upload result with CID and gateway URL
   */
  async uploadFile(fileBuffer, fileName, metadata = {}) {
    if (!this.initialized) {
      throw new Error('IPFS Service not initialized');
    }

    const uploadId = uuidv4();
    logger.info('Starting IPFS upload', { uploadId, fileName, size: fileBuffer.length });

    try {
      if (this.provider === 'pinata') {
        return await this.uploadToPinata(fileBuffer, fileName, metadata, uploadId);
      } else {
        throw new Error(`Unsupported IPFS provider: ${this.provider}`);
      }
    } catch (error) {
      logger.error('IPFS upload failed', { uploadId, error: error.message });
      throw error;
    }
  }

  /**
   * Upload to Pinata
   */
  async uploadToPinata(fileBuffer, fileName, metadata, uploadId) {
    const formData = new FormData();
    formData.append('file', fileBuffer, { filename: fileName });

    // Add metadata
    const pinataMetadata = JSON.stringify({
      name: `grampulse-proof-${uploadId}`,
      keyvalues: {
        app: 'grampulse',
        type: 'proof-of-resolution',
        uploadId,
        ...metadata,
      },
    });
    formData.append('pinataMetadata', pinataMetadata);

    // Pin options
    const pinataOptions = JSON.stringify({
      cidVersion: 1,
    });
    formData.append('pinataOptions', pinataOptions);

    const response = await axios.post(
      `${IPFS_CONFIG.pinata.apiUrl}/pinning/pinFileToIPFS`,
      formData,
      {
        maxBodyLength: Infinity,
        headers: {
          'Content-Type': `multipart/form-data; boundary=${formData._boundary}`,
          'pinata_api_key': this.apiKey,
          'pinata_secret_api_key': this.apiSecret,
        },
      }
    );

    const { IpfsHash, PinSize, Timestamp } = response.data;

    logger.info('IPFS upload successful', {
      uploadId,
      cid: IpfsHash,
      size: PinSize,
    });

    return {
      success: true,
      cid: IpfsHash,
      size: PinSize,
      timestamp: Timestamp,
      gatewayUrl: `${IPFS_CONFIG.pinata.gateway}/${IpfsHash}`,
      uploadId,
    };
  }

  /**
   * Upload JSON data to IPFS
   * 
   * @param {Object} data - JSON data to upload
   * @param {string} name - Name for the JSON file
   * @returns {Object} - Upload result with CID
   */
  async uploadJSON(data, name = 'data.json') {
    if (!this.initialized) {
      throw new Error('IPFS Service not initialized');
    }

    const uploadId = uuidv4();
    logger.info('Starting IPFS JSON upload', { uploadId, name });

    try {
      if (this.provider === 'pinata') {
        return await this.uploadJSONToPinata(data, name, uploadId);
      } else {
        throw new Error(`Unsupported IPFS provider: ${this.provider}`);
      }
    } catch (error) {
      logger.error('IPFS JSON upload failed', { uploadId, error: error.message });
      throw error;
    }
  }

  /**
   * Upload JSON to Pinata
   */
  async uploadJSONToPinata(data, name, uploadId) {
    const response = await axios.post(
      `${IPFS_CONFIG.pinata.apiUrl}/pinning/pinJSONToIPFS`,
      {
        pinataContent: data,
        pinataMetadata: {
          name: `grampulse-${name}-${uploadId}`,
          keyvalues: {
            app: 'grampulse',
            type: 'json-data',
            uploadId,
          },
        },
        pinataOptions: {
          cidVersion: 1,
        },
      },
      {
        headers: {
          'Content-Type': 'application/json',
          'pinata_api_key': this.apiKey,
          'pinata_secret_api_key': this.apiSecret,
        },
      }
    );

    const { IpfsHash, PinSize, Timestamp } = response.data;

    logger.info('IPFS JSON upload successful', {
      uploadId,
      cid: IpfsHash,
    });

    return {
      success: true,
      cid: IpfsHash,
      size: PinSize,
      timestamp: Timestamp,
      gatewayUrl: `${IPFS_CONFIG.pinata.gateway}/${IpfsHash}`,
      uploadId,
    };
  }

  /**
   * Create a proof-of-resolution package
   * 
   * Bundles proof media with metadata into a single IPFS upload
   * 
   * @param {Object} params - Proof parameters
   * @param {string} params.grievanceId - Grievance document ID
   * @param {string} params.villageId - Village identifier
   * @param {string} params.resolverRole - Role of resolver
   * @param {string} params.resolverId - User ID of resolver
   * @param {string} params.description - Resolution description
   * @param {Array} params.mediaFiles - Array of {buffer, fileName, mimeType}
   * @returns {Object} - Package upload result
   */
  async createProofPackage(params) {
    const {
      grievanceId,
      villageId,
      resolverRole,
      resolverId,
      description,
      mediaFiles = [],
    } = params;

    const uploadId = uuidv4();
    logger.info('Creating proof package', { uploadId, grievanceId, mediaCount: mediaFiles.length });

    try {
      // Upload media files first
      const mediaUploads = [];
      for (const media of mediaFiles) {
        const result = await this.uploadFile(media.buffer, media.fileName, {
          grievanceId,
          mediaType: media.mimeType,
        });
        mediaUploads.push({
          cid: result.cid,
          fileName: media.fileName,
          mimeType: media.mimeType,
          size: result.size,
          gatewayUrl: result.gatewayUrl,
        });
      }

      // Create metadata JSON
      const proofMetadata = {
        version: '1.0.0',
        type: 'grampulse-proof-of-resolution',
        grievanceId,
        villageId,
        resolution: {
          resolverRole,
          resolverId,
          description,
          timestamp: new Date().toISOString(),
          unixTimestamp: Math.floor(Date.now() / 1000),
        },
        media: mediaUploads,
        packageId: uploadId,
      };

      // Upload metadata JSON
      const metadataResult = await this.uploadJSON(
        proofMetadata,
        `proof-${grievanceId}`
      );

      logger.info('Proof package created', {
        uploadId,
        grievanceId,
        packageCid: metadataResult.cid,
        mediaCount: mediaUploads.length,
      });

      return {
        success: true,
        packageCid: metadataResult.cid,
        packageUrl: metadataResult.gatewayUrl,
        metadata: proofMetadata,
        mediaFiles: mediaUploads,
        uploadId,
      };

    } catch (error) {
      logger.error('Failed to create proof package', {
        uploadId,
        grievanceId,
        error: error.message,
      });
      throw error;
    }
  }

  /**
   * Get content from IPFS
   * 
   * @param {string} cid - IPFS CID
   * @returns {Object} - Content data
   */
  async getContent(cid) {
    const gateway = this.provider === 'pinata' 
      ? IPFS_CONFIG.pinata.gateway 
      : IPFS_CONFIG.web3Storage.gateway;

    try {
      const response = await axios.get(`${gateway}/${cid}`, {
        timeout: 30000,
      });
      return response.data;
    } catch (error) {
      logger.error('Failed to fetch IPFS content', { cid, error: error.message });
      throw error;
    }
  }

  /**
   * Get gateway URL for a CID
   */
  getGatewayUrl(cid) {
    const gateway = this.provider === 'pinata' 
      ? IPFS_CONFIG.pinata.gateway 
      : IPFS_CONFIG.web3Storage.gateway;
    return `${gateway}/${cid}`;
  }

  /**
   * Check if service is ready
   */
  isReady() {
    return this.initialized;
  }
}

// Singleton instance
const ipfsService = new IPFSService();

module.exports = ipfsService;
